### DISCUSSION TITLE
How to use hub.read_line() Without wait in pybricksdev?

### DISCUSSION THREAD

**Discussion by johnscary-ev3** [pybricks/support#1190] [Q&A]
I am trying to read data sent from hub into my program for analysis using prybricksdev package.
I would like to check for data from the hub while also getting keyboard inputs to send to the hub.
The hub.read_line() call will get lines of data from the hub but will wait if nothing is available.
I would like to Not wait but just move on if no data and check the keyboard.
Is there a way to do this?  Check if data available first somehow before requesting it?

I tried to use the example code below using a timeout approach for the read_line() calls.
This actually works for a while during my hub initialization when it is sending lines of data.
After that it gets intermittent giving me timeout messages and ignoring the hub data being sent.
Then I get some error messages and timeouts mixed together and no hub data. A sample output is below.

Is there a way to do this?
Thanks

```
async def read_key():
   
    if platform.system() == 'Windows':
        while True:
            #Check if keyboard character was hit
            if msvcrt.kbhit():
                return msvcrt.getch()
           
            sleep(.250)
            # Check for data from Hub while waiting for keyboard character
            try:
                line = await asyncio.wait_for(hub.read_line(), .1)
                # Process Data from Hub
                print("read_key: print_line=", line)
            except asyncio.TimeoutError:
                print("read_key: read_line TimeoutError")
```

Errors from my program:
```
read_key: read_line TimeoutError
read_key: read_line TimeoutError
read_key: read_line TimeoutError
Task was destroyed but it is pending!
task: <Task pending name='Task-78' coro=<Event.wait() done, defined at C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\asyncio\locks.py:201> wait_for=<Future pending cb=[Task.task_wakeup()]>>
Task was destroyed but it is pending!
task: <Task pending name='Task-81' coro=<Event.wait() done, defined at C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.10_3.10.3056.0_x64__qbz5n2kfra8p0\lib\asyncio\locks.py:201> wait_for=<Future pending cb=[Task.task_wakeup()]>>
```

**Comment by dlech** [comment]
There isn't really a good way to get input while using asyncio using only standard Python libraries. I would suggest using something like https://pypi.org/project/aioconsole/.

Calling blocking functions like `sleep()` in asyncio is going to break things. You can call `await asyncio.sleep()` instead if you really need to. But, if you are using it in a loop to poll for something, then there is almost certainly a better way wait for that event.

If you are using Python >= 3.11 and you need to do multiple things in parallel, I would recommend using [asyncio.TaskGroup](https://docs.python.org/3/library/asyncio-task.html#asyncio.TaskGroup). Then you can wait for terminal input in one task and wait for `read_line()` in another task.

**Comment by johnscary-ev3** [comment]
@dlech 
Thanks for reply and suggestions.
However, I think this is getting overly complicated.
What I think is really needed is an easy way for hub.read_line() to let us know if it has a line to read or gives an exception or just a null string.
So, I could write this code below and not have it hang when no line is available.
The while loop at the end of code below allows me to send one command to hub and get the response, but then hangs.
Thanks


```python
async def forwarder(hub: PybricksHub):
    print("forwarder: Start")

    queue = asyncio.Queue()

    # wait for user program on the hub to start
    with hub.status_observable.subscribe(lambda s: queue.put_nowait(s)):
        while True:
            status = await queue.get()

            if status & StatusFlag.USER_PROGRAM_RUNNING:
                break

    print('forwarder: Hub Running')

    loop = asyncio.get_running_loop()
 
    # Keyboard command input loop
    while True:
        command = await loop.run_in_executor(None, read_key)
        print('forwarder: command =', command)

        # Try to send to the receiving hub
        try:
            await hub.write(bytes([ord(command)]))
        except:
            pass

        # Check for data from Hub
        while True:
            try:
                print("forwarder: call read_line")
                data = await hub.read_line()
                print("forwarder: data=", data)
                if data == None:
                    break
            except:
                print("forwarder: read_line Error")
                break

```

**Reply by dlech** [reply]
> However, I think this is getting overly complicated.

For robust communications, I think it is really best to have separate tasks to avoid race conditions. There needs to be some sort of synchronization/handshake between the hub and the computer to wait for a "ready to receive a command" signal from the hub before sending a command to avoid losing commands. And at the same time, the computer needs to be queuing key presses while waiting from the ready signal from the hub in order to avoid losing key presses.

**Comment by johnscary-ev3** [comment]
Yes, I see your point about handshakes.
For now, however, I dug into the code of how you all implemented hub.read_line() and realized that it uses Python Queues.
Python Queues has an "empty" queue call, so I tried that.  "hub._stdout_line_queue.empty()"
So first I check if any keyboard commands are available and then use this call to check for line data available as below.
This works ok with no errors, however the line data could be a left over from previous command, so I need to check what I get.

The reason to do all this is that in addition to logging command responses on the terminal, I hope to have the robot send heading and location data after any moves I request. Then I intend to display that on a plot using "vispy" plotting package. This is just the first step.

Thanks.

```python
# Keyboard command input loop
    while True:
        command = await loop.run_in_executor(None, read_key)
        if command != None:
            print('forwarder: command =', command)

            # Try to send to the receiving hub
            try:
                await hub.write(bytes([ord(command)]))
            except:
                pass

        # Check for data from Hub
        while True:
            try:
                if not hub._stdout_line_queue.empty():
                    data = await hub.read_line()
                    print("forwarder: hub.read_line data=", data)
                else:
                    break
            except:
                print("forwarder: hub.read_line Error")
                break
```
